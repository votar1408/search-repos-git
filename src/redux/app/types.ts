export const LOADING: string = 'LOADING';

export interface AppLoadingAction {
    type: typeof LOADING;
    payload: boolean;
}

export type AppAction = AppLoadingAction;

export interface AppState {
    loading: boolean;
}
