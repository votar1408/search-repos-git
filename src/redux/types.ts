import {AppState} from './app/types';

export interface RootAppState {
    app: AppState;
}

interface RootState extends RootAppState {}

export const selectApp = (state: RootState) => state.app;
