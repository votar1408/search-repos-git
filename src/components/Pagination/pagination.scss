@import '../../assets/scss/variable.scss';

.pagination {
    display: flex;
    align-items: center;
    justify-content: center;
    margin: 18px 0;
}
.pagination__button {
    width: 40px;
    height: 36px;
    margin: 0px 6px;
    border: 2px solid $color-main;
    background: $color-main;
    text-align: center;
    color: #fff;
    border-radius: 5px;
    cursor: pointer;
    font-size: 24px;
    transition: 0.5s;
    &:hover {
        background: lighten($color: $color-main, $amount: 10);
    }
    &:disabled {
        cursor: default;
        background: lighten($color: $color-light-grey, $amount: 10);
        border: 2px solid $color-grey;
    }
}
