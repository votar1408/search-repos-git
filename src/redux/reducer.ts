import {combineReducers} from 'redux';

import {appReducer} from './app/reducer';

export const reducer = combineReducers({
    app: appReducer
});
