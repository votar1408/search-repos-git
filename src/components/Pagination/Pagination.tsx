import './pagination.scss';

import React, {
    FC,
    ReactElement,
    useCallback,
    useMemo
} from 'react';

import {
    useDispatch,
    useSelector
} from 'react-redux';

import {common} from '../../consts';
import {RepositoryServerInfo} from '../../interfaces/ServerData';
import {changePageAction} from '../../redux/app/actions';
import {selectApp} from '../../redux/types';

export const Pagination: FC = () => {
    const {page, repositories, searchValue, loading} = useSelector(selectApp);
    const dispatch = useDispatch();

    const listForPage: RepositoryServerInfo[] = !repositories[page] ? repositories[page - 1] : repositories[page];

    const onChangePageHandler = useCallback(
        (value: number) => {
            dispatch(changePageAction(value));
        },
        [dispatch]
    );

    const buttons: ReactElement | null = useMemo(() => {
        return searchValue && repositories.length > 0 ? (
            <div className="pagination">
                <button
                    type="submit"
                    className="pagination__button"
                    disabled={page === common.getDefaultPage || loading}
                    data-id="left-button"
                    onClick={() => onChangePageHandler(-1)}
                >
                    <i className="fa fa-angle-left"></i>
                </button>
                <button
                    type="submit"
                    className="pagination__button"
                    disabled={!listForPage?.length || loading}
                    data-id="right-button"
                    onClick={() => onChangePageHandler(1)}
                >
                    <i className="fa fa-angle-right"></i>
                </button>
            </div>
        ) : null;
    }, [listForPage, loading, onChangePageHandler, page, repositories.length, searchValue]);

    return <>{buttons}</>;
};
