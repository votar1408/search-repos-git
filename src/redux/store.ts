/* eslint-disable no-underscore-dangle */
import {
    applyMiddleware,
    compose,
    createStore,
    Middleware
} from 'redux';
import thunk from 'redux-thunk';

import {reducer} from './reducer';

const middleware: [Middleware] = [thunk];

export const store = createStore(reducer, compose(applyMiddleware(...middleware)));
