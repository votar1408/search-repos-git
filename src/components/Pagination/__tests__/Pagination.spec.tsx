import React from 'react';

import {mount} from 'enzyme';
import {shallowToJson} from 'enzyme-to-json';
import {Provider} from 'react-redux';
import {BrowserRouter as Router} from 'react-router-dom';
import createMockStore, {
    MockStoreCreator,
    MockStoreEnhanced
} from 'redux-mock-store';

import {RootAppState} from '../../../redux/types';
import {Pagination} from '../Pagination';

describe('Pagination', () => {
    const getInitialState: any = (pageValue: number = 1): RootAppState => {
        return {
            app: {
                loading: false,
                repositories: [
                    undefined,
                    [
                        {
                            id: 1,
                            name: 'test1',
                            description: 'some description 1',
                            owner: {
                                login: 'login1',
                                avatar_url: 'some url1'
                            }
                        },
                        {
                            id: 2,
                            name: 'test2',
                            description: 'some description 2',
                            owner: {
                                login: 'login2',
                                avatar_url: 'some url2'
                            }
                        }
                    ]
                ],
                page: pageValue,
                searchValue: 'test'
            }
        };
    };

    const mockStore: MockStoreCreator = createMockStore();

    let store: MockStoreEnhanced;
    let wrapper: any;

    const getWrapper = (state: RootAppState) => {
        store = mockStore(state);
        return mount(
            <Router>
                <Provider store={store}>
                    <Pagination />
                </Provider>
            </Router>
        );
    };

    beforeEach(() => {
        wrapper = getWrapper(getInitialState());
    });

    it('should render correctly', () => {
        expect(shallowToJson(wrapper)).toMatchSnapshot();
    });

    it('renders without crashing', () => {
        expect(wrapper.find(Pagination).length).toEqual(1);
    });

    it('should have 2 buttons', () => {
        expect(wrapper.find('button').length).toEqual(2);
    });

    it('left button should be disabled on first page', () => {
        let leftButton = wrapper.find('button[data-id="left-button"]');
        let rightButton = wrapper.find('button[data-id="right-button"]');
        expect(leftButton.prop('disabled')).toEqual(true);
        expect(rightButton.prop('disabled')).toEqual(false);
    });

    it('should update disabled state after next click', () => {
        wrapper = getWrapper(getInitialState(2));

        let leftButton = wrapper.find('button[data-id="left-button"]');
        let rightButton = wrapper.find('button[data-id="right-button"]');
        expect(leftButton.prop('disabled')).toEqual(false);
        expect(rightButton.prop('disabled')).toEqual(false);

        rightButton.simulate('click');

        leftButton = wrapper.find('button[data-id="left-button"]');
        rightButton = wrapper.find('button[data-id="right-button"]');
        expect(leftButton.prop('disabled')).toEqual(false);
        expect(rightButton.prop('disabled')).toEqual(false);
    });
});
