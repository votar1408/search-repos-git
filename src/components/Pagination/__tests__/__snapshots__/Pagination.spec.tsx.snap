// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination should render correctly 1`] = `
<BrowserRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Provider
      store={
        Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        }
      }
    >
      <Pagination>
        <div
          className="pagination"
        >
          <button
            className="pagination__button"
            data-id="left-button"
            disabled={true}
            onClick={[Function]}
            type="submit"
          >
            <i
              className="fa fa-angle-left"
            />
          </button>
          <button
            className="pagination__button"
            data-id="right-button"
            disabled={false}
            onClick={[Function]}
            type="submit"
          >
            <i
              className="fa fa-angle-right"
            />
          </button>
        </div>
      </Pagination>
    </Provider>
  </Router>
</BrowserRouter>
`;
